File: backend/app/schemas/audit.py
"""
Audit Log Schemas

Pydantic schemas for audit logging API.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class ActionCategory(str, Enum):
    """Categories of audit actions"""
    AUTH = "auth"
    DASHBOARD = "dashboard"
    CHART = "chart"
    CONNECTION = "connection"
    QUERY = "query"
    DATA = "data"
    WORKSPACE = "workspace"
    ADMIN = "admin"
    SYSTEM = "system"


class ActionType(str, Enum):
    """Types of actions"""
    VIEW = "view"
    CREATE = "create"
    UPDATE = "update"
    DELETE = "delete"
    EXPORT = "export"
    SHARE = "share"
    EXECUTE = "execute"
    LOGIN = "login"
    LOGOUT = "logout"


class AlertSeverity(str, Enum):
    """Security alert severity levels"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class AlertStatus(str, Enum):
    """Security alert status"""
    OPEN = "open"
    INVESTIGATING = "investigating"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"


# Audit Log Schemas

class AuditLogBase(BaseModel):
    """Base audit log schema"""
    action: str
    action_category: Optional[ActionCategory] = None
    action_type: Optional[ActionType] = None
    resource_type: Optional[str] = None
    resource_id: Optional[str] = None
    resource_name: Optional[str] = None


class AuditLogCreate(AuditLogBase):
    """Schema for creating an audit log entry"""
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    session_id: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    request_method: Optional[str] = None
    request_path: Optional[str] = None
    request_body: Optional[dict[str, Any]] = None
    response_status: Optional[int] = None
    duration_ms: Optional[int] = None
    workspace_id: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    success: bool = True
    error_message: Optional[str] = None


class AuditLogResponse(AuditLogBase):
    """Schema for audit log response"""
    id: str
    timestamp: datetime
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    user_name: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    request_method: Optional[str] = None
    request_path: Optional[str] = None
    response_status: Optional[int] = None
    duration_ms: Optional[int] = None
    workspace_id: Optional[str] = None
    success: int = 1
    error_message: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    class Config:
        from_attributes = True


class AuditLogFilter(BaseModel):
    """Schema for filtering audit logs"""
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    action: Optional[str] = None
    action_category: Optional[ActionCategory] = None
    action_type: Optional[ActionType] = None
    resource_type: Optional[str] = None
    resource_id: Optional[str] = None
    workspace_id: Optional[str] = None
    ip_address: Optional[str] = None
    success_only: Optional[bool] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class AuditLogSummary(BaseModel):
    """Summary of audit logs"""
    total_events: int
    successful_events: int
    failed_events: int
    unique_users: int
    top_actions: list[dict[str, Any]]
    activity_by_hour: list[dict[str, Any]]
    activity_by_category: dict[str, int]


class AuditLogExport(BaseModel):
    """Schema for audit log export"""
    format: str = "csv"  # csv, json, xlsx
    filters: Optional[AuditLogFilter] = None
    include_metadata: bool = False
    max_records: int = Field(default=10000, le=100000)


# Security Alert Schemas

class SecurityAlertBase(BaseModel):
    """Base security alert schema"""
    alert_type: str
    severity: AlertSeverity
    title: str
    description: Optional[str] = None


class SecurityAlertCreate(SecurityAlertBase):
    """Schema for creating a security alert"""
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    ip_address: Optional[str] = None
    related_audit_ids: list[str] = Field(default_factory=list)
    workspace_id: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class SecurityAlertResponse(SecurityAlertBase):
    """Schema for security alert response"""
    id: str
    created_at: datetime
    resolved_at: Optional[datetime] = None
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    ip_address: Optional[str] = None
    related_audit_ids: list[str] = Field(default_factory=list)
    workspace_id: Optional[str] = None
    status: AlertStatus
    resolved_by: Optional[str] = None
    resolution_notes: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    class Config:
        from_attributes = True


class SecurityAlertUpdate(BaseModel):
    """Schema for updating a security alert"""
    status: Optional[AlertStatus] = None
    resolution_notes: Optional[str] = None


class SecurityAlertFilter(BaseModel):
    """Schema for filtering security alerts"""
    alert_type: Optional[str] = None
    severity: Optional[AlertSeverity] = None
    status: Optional[AlertStatus] = None
    user_id: Optional[str] = None
    workspace_id: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


# Activity Timeline

class ActivityTimelineEntry(BaseModel):
    """A single entry in the activity timeline"""
    timestamp: datetime
    action: str
    description: str
    user_name: Optional[str] = None
    user_email: Optional[str] = None
    resource_type: Optional[str] = None
    resource_name: Optional[str] = None
    success: bool = True
    icon: Optional[str] = None
    color: Optional[str] = None


class UserActivitySummary(BaseModel):
    """Summary of user activity"""
    user_id: str
    user_email: Optional[str] = None
    user_name: Optional[str] = None
    total_actions: int
    last_active: Optional[datetime] = None
    most_used_features: list[dict[str, Any]]
    login_count: int
    failed_login_count: int


# Admin Dashboard

class AuditDashboardStats(BaseModel):
    """Statistics for audit dashboard"""
    total_events_today: int
    total_events_week: int
    active_users_today: int
    failed_logins_today: int
    open_alerts: int
    critical_alerts: int
    top_users: list[dict[str, Any]]
    recent_alerts: list[SecurityAlertResponse]


# Constants

ACTION_ICONS = {
    "auth.login": "log-in",
    "auth.logout": "log-out",
    "auth.login_failed": "x-circle",
    "dashboard.view": "eye",
    "dashboard.create": "plus",
    "dashboard.update": "edit",
    "dashboard.delete": "trash",
    "dashboard.share": "share",
    "chart.view": "eye",
    "chart.create": "plus",
    "chart.update": "edit",
    "chart.delete": "trash",
    "chart.export": "download",
    "connection.create": "database",
    "connection.test": "refresh",
    "query.execute": "play",
    "data.export": "download",
    "workspace.create": "folder-plus",
    "workspace.member_add": "user-plus",
    "workspace.member_remove": "user-minus",
}

ACTION_COLORS = {
    "auth.login": "green",
    "auth.logout": "gray",
    "auth.login_failed": "red",
    "dashboard.create": "blue",
    "dashboard.update": "yellow",
    "dashboard.delete": "red",
    "chart.create": "blue",
    "chart.delete": "red",
    "chart.export": "purple",
    "data.export": "purple",
}

SENSITIVE_FIELDS = [
    "password",
    "secret",
    "token",
    "api_key",
    "credentials",
    "private_key",
    "access_token",
    "refresh_token",
]


def sanitize_request_body(body: dict) -> dict:
    """Remove sensitive fields from request body"""
    if not body:
        return {}

    sanitized = {}
    for key, value in body.items():
        key_lower = key.lower()
        if any(sensitive in key_lower for sensitive in SENSITIVE_FIELDS):
            sanitized[key] = "[REDACTED]"
        elif isinstance(value, dict):
            sanitized[key] = sanitize_request_body(value)
        else:
            sanitized[key] = value

    return sanitized


def format_action_description(action: str, resource_type: str = None, resource_name: str = None) -> str:
    """Generate human-readable description for an action"""
    action_parts = action.split(".")
    if len(action_parts) != 2:
        return action

    category, action_type = action_parts

    descriptions = {
        ("auth", "login"): "logged in",
        ("auth", "logout"): "logged out",
        ("auth", "login_failed"): "failed to log in",
        ("dashboard", "view"): f"viewed dashboard{f' {resource_name}' if resource_name else ''}",
        ("dashboard", "create"): f"created dashboard{f' {resource_name}' if resource_name else ''}",
        ("dashboard", "update"): f"updated dashboard{f' {resource_name}' if resource_name else ''}",
        ("dashboard", "delete"): f"deleted dashboard{f' {resource_name}' if resource_name else ''}",
        ("dashboard", "share"): f"shared dashboard{f' {resource_name}' if resource_name else ''}",
        ("chart", "view"): f"viewed chart{f' {resource_name}' if resource_name else ''}",
        ("chart", "create"): f"created chart{f' {resource_name}' if resource_name else ''}",
        ("chart", "update"): f"updated chart{f' {resource_name}' if resource_name else ''}",
        ("chart", "delete"): f"deleted chart{f' {resource_name}' if resource_name else ''}",
        ("chart", "export"): f"exported chart{f' {resource_name}' if resource_name else ''}",
        ("connection", "create"): f"created connection{f' {resource_name}' if resource_name else ''}",
        ("connection", "test"): f"tested connection{f' {resource_name}' if resource_name else ''}",
        ("query", "execute"): "executed a query",
        ("data", "export"): "exported data",
        ("workspace", "create"): f"created workspace{f' {resource_name}' if resource_name else ''}",
        ("workspace", "member_add"): "added a member to workspace",
        ("workspace", "member_remove"): "removed a member from workspace",
    }

    return descriptions.get((category, action_type), action)

File: frontend/src/types/audit.ts
/**
 * Audit Log Types
 *
 * TypeScript types for audit logging and security alerts.
 */

// Enums

export type ActionCategory =
  | 'auth'
  | 'dashboard'
  | 'chart'
  | 'connection'
  | 'query'
  | 'data'
  | 'workspace'
  | 'admin'
  | 'system';

export type ActionType =
  | 'view'
  | 'create'
  | 'update'
  | 'delete'
  | 'export'
  | 'share'
  | 'execute'
  | 'login'
  | 'logout';

export type AlertSeverity = 'low' | 'medium' | 'high' | 'critical';

export type AlertStatus = 'open' | 'investigating' | 'resolved' | 'dismissed';

// Audit Log Types

export interface AuditLog {
  id: string;
  timestamp: string;
  user_id?: string;
  user_email?: string;
  user_name?: string;
  action: string;
  action_category?: ActionCategory;
  action_type?: ActionType;
  resource_type?: string;
  resource_id?: string;
  resource_name?: string;
  ip_address?: string;
  user_agent?: string;
  request_method?: string;
  request_path?: string;
  response_status?: number;
  duration_ms?: number;
  workspace_id?: string;
  success: number;
  error_message?: string;
  metadata: Record<string, unknown>;
}

export interface AuditLogFilter {
  user_id?: string;
  user_email?: string;
  action?: string;
  action_category?: ActionCategory;
  action_type?: ActionType;
  resource_type?: string;
  resource_id?: string;
  workspace_id?: string;
  ip_address?: string;
  success_only?: boolean;
  start_date?: string;
  end_date?: string;
}

export interface AuditLogSummary {
  total_events: number;
  successful_events: number;
  failed_events: number;
  unique_users: number;
  top_actions: Array<{ action: string; count: number }>;
  activity_by_hour: Array<{ hour: number; count: number }>;
  activity_by_category: Record<string, number>;
}

export interface AuditLogExport {
  format: 'csv' | 'json' | 'xlsx';
  filters?: AuditLogFilter;
  include_metadata?: boolean;
  max_records?: number;
}

// Security Alert Types

export interface SecurityAlert {
  id: string;
  created_at: string;
  resolved_at?: string;
  alert_type: string;
  severity: AlertSeverity;
  title: string;
  description?: string;
  user_id?: string;
  user_email?: string;
  ip_address?: string;
  related_audit_ids: string[];
  workspace_id?: string;
  status: AlertStatus;
  resolved_by?: string;
  resolution_notes?: string;
  metadata: Record<string, unknown>;
}

export interface SecurityAlertCreate {
  alert_type: string;
  severity: AlertSeverity;
  title: string;
  description?: string;
  user_id?: string;
  user_email?: string;
  ip_address?: string;
  related_audit_ids?: string[];
  workspace_id?: string;
  metadata?: Record<string, unknown>;
}

export interface SecurityAlertUpdate {
  status?: AlertStatus;
  resolution_notes?: string;
}

export interface SecurityAlertFilter {
  alert_type?: string;
  severity?: AlertSeverity;
  status?: AlertStatus;
  user_id?: string;
  workspace_id?: string;
  start_date?: string;
  end_date?: string;
}

// Activity Types

export interface ActivityTimelineEntry {
  timestamp: string;
  action: string;
  description: string;
  user_name?: string;
  user_email?: string;
  resource_type?: string;
  resource_name?: string;
  success: boolean;
  icon?: string;
  color?: string;
}

export interface UserActivitySummary {
  user_id: string;
  user_email?: string;
  user_name?: string;
  total_actions: number;
  last_active?: string;
  most_used_features: Array<{ feature: string; count: number }>;
  login_count: number;
  failed_login_count: number;
}

// Dashboard Stats

export interface AuditDashboardStats {
  total_events_today: number;
  total_events_week: number;
  active_users_today: number;
  failed_logins_today: number;
  open_alerts: number;
  critical_alerts: number;
  top_users: Array<{ user_email: string; action_count: number }>;
  recent_alerts: SecurityAlert[];
}

// Anomaly Detection

export interface AnomalyCheckResult {
  checked_at: string;
  anomalies_found: number;
  alerts_created: string[];
}

// Archive

export interface ArchiveResult {
  archived_count: number;
  days_kept: number;
  archived_at: string;
}

// Constants

export const ACTION_CATEGORY_LABELS: Record<ActionCategory, string> = {
  auth: 'Authentication',
  dashboard: 'Dashboard',
  chart: 'Chart',
  connection: 'Connection',
  query: 'Query',
  data: 'Data',
  workspace: 'Workspace',
  admin: 'Admin',
  system: 'System',
};

export const ACTION_TYPE_LABELS: Record<ActionType, string> = {
  view: 'View',
  create: 'Create',
  update: 'Update',
  delete: 'Delete',
  export: 'Export',
  share: 'Share',
  execute: 'Execute',
  login: 'Login',
  logout: 'Logout',
};

export const ALERT_SEVERITY_LABELS: Record<AlertSeverity, string> = {
  low: 'Low',
  medium: 'Medium',
  high: 'High',
  critical: 'Critical',
};

export const ALERT_SEVERITY_COLORS: Record<AlertSeverity, string> = {
  low: 'blue',
  medium: 'yellow',
  high: 'orange',
  critical: 'red',
};

export const ALERT_STATUS_LABELS: Record<AlertStatus, string> = {
  open: 'Open',
  investigating: 'Investigating',
  resolved: 'Resolved',
  dismissed: 'Dismissed',
};

export const ALERT_STATUS_COLORS: Record<AlertStatus, string> = {
  open: 'red',
  investigating: 'yellow',
  resolved: 'green',
  dismissed: 'gray',
};

export const ACTION_ICONS: Record<string, string> = {
  'auth.login': 'log-in',
  'auth.logout': 'log-out',
  'auth.login_failed': 'x-circle',
  'dashboard.view': 'eye',
  'dashboard.create': 'plus',
  'dashboard.update': 'edit',
  'dashboard.delete': 'trash',
  'dashboard.share': 'share',
  'chart.view': 'eye',
  'chart.create': 'plus',
  'chart.update': 'edit',
  'chart.delete': 'trash',
  'chart.export': 'download',
  'connection.create': 'database',
  'connection.test': 'refresh-cw',
  'query.execute': 'play',
  'data.export': 'download',
  'workspace.create': 'folder-plus',
  'workspace.member_add': 'user-plus',
  'workspace.member_remove': 'user-minus',
};

export const ACTION_COLORS: Record<string, string> = {
  'auth.login': 'green',
  'auth.logout': 'gray',
  'auth.login_failed': 'red',
  'dashboard.create': 'blue',
  'dashboard.update': 'yellow',
  'dashboard.delete': 'red',
  'chart.create': 'blue',
  'chart.delete': 'red',
  'chart.export': 'purple',
  'data.export': 'purple',
};

// Helper Functions

export function getActionIcon(action: string): string {
  return ACTION_ICONS[action] || 'activity';
}

export function getActionColor(action: string): string {
  return ACTION_COLORS[action] || 'gray';
}

export function getSeverityColor(severity: AlertSeverity): string {
  return ALERT_SEVERITY_COLORS[severity];
}

export function getStatusColor(status: AlertStatus): string {
  return ALERT_STATUS_COLORS[status];
}

export function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  if (ms < 60000) return `${(ms / 1000).toFixed(1)}s`;
  return `${Math.floor(ms / 60000)}m ${Math.floor((ms % 60000) / 1000)}s`;
}

export function formatTimestamp(timestamp: string): string {
  const date = new Date(timestamp);
  return date.toLocaleString(undefined, {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  });
}

export function formatRelativeTime(timestamp: string): string {
  const date = new Date(timestamp);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffSeconds = Math.floor(diffMs / 1000);
  const diffMinutes = Math.floor(diffSeconds / 60);
  const diffHours = Math.floor(diffMinutes / 60);
  const diffDays = Math.floor(diffHours / 24);

  if (diffSeconds < 60) return 'Just now';
  if (diffMinutes < 60) return `${diffMinutes}m ago`;
  if (diffHours < 24) return `${diffHours}h ago`;
  if (diffDays < 7) return `${diffDays}d ago`;
  return date.toLocaleDateString();
}

export function isHighSeverity(severity: AlertSeverity): boolean {
  return severity === 'high' || severity === 'critical';
}

export function isOpenAlert(status: AlertStatus): boolean {
  return status === 'open' || status === 'investigating';
}
